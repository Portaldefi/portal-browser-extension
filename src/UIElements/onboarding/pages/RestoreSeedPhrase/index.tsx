import React, { useCallback, useMemo, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { Header, Message, Button, Grid, Segment, Icon, Form } from 'semantic-ui-react';
import { useForm, NestedValue } from 'react-hook-form';
import { isEqual } from 'lodash';

import { useAppSelector } from '../../hooks';

type FormValue = {
  phrase1: string, phrase2: string, phrase3: string,
  phrase4: string, phrase5: string, phrase6: string,
  phrase7: string, phrase8: string, phrase9: string,
  phrase10: string, phrase11: string, phrase12: string
};

const defaultPhrases = {
  phrase1: '', phrase2: '', phrase3: '', 
  phrase4: '', phrase5: '', phrase6: '',
  phrase7: '', phrase8: '', phrase9: '',
  phrase10: '', phrase11: '', phrase12: '',
}

export default () => {
  const [ isDirty, setIsDirty ] = useState(false);
  const navigate = useNavigate();
  const phrase = useAppSelector(state => state.phrase);
  const { setValue, getValues, watch } = useForm<FormValue>({ defaultValues: defaultPhrases });

  const phrases = useMemo(() => {
    return getValues()
  }, [
    watch('phrase1'), watch('phrase2'), watch('phrase3'),
    watch('phrase4'), watch('phrase5'), watch('phrase6'),
    watch('phrase7'), watch('phrase8'), watch('phrase9'),
    watch('phrase10'), watch('phrase11'), watch('phrase12'),
  ]);
  const isCorrectPhrase = useMemo(() => isEqual(Object.keys(phrases).map(key => phrases[key]), phrase.SRF_List), [phrases, phrase.SRF_List]);

  const handleBack = useCallback(() => {
    navigate('/seed-phrase');
  }, []);

  const handleConfirm = useCallback(() => {
    if (isCorrectPhrase) {
      navigate('/congrats', { state: { mode: 'create' } });
    } else {
      setIsDirty(true);
    }
  }, [isCorrectPhrase]);

  return (
    <Segment className='board'>
      <Grid>
        <Grid.Row centered>
          <Header size='medium' className='heading'>Restore Seed Phrase</Header>
        </Grid.Row>
        <Grid.Row centered>
          <Form className='form-container'>
            {[...new Array(Math.ceil(phrase.SRF_Length / 3))].map((_, rowIdx) => (
              <Form.Group inline widths={3} key={rowIdx}>
                {[...new Array(3)].map((_, colIdx) => {
                  const id = rowIdx * 3 + colIdx + 1;

                  return <Form.Input label={`${id}.`} width={16} key={colIdx} value={phrases[`phrase${id}`]} onChange={(e) => { setValue(`phrase${id}`, e.target.value); setIsDirty(true); }} />
                })}
              </Form.Group>
            ))}
          </Form>
        </Grid.Row>
        <Grid.Row>
          <Header as='p' size='small' color='grey' className='description extra-former-blank extra-latter-blank'>
            WARNING: Improper backup of the seed phrase will result in loss of funds!
          </Header>
        </Grid.Row>
        {isDirty && !isCorrectPhrase && <Grid.Row>
          <Message warning>
            Invalid Secret Recovery Phrase
          </Message>
        </Grid.Row>}
        <Grid.Row centered columns={2}>
          <Grid.Column>
            <Button className='blank-button' onClick={handleBack}><Icon name='angle left' />Back</Button>
          </Grid.Column>
          <Grid.Column>
            <Button className='primary-button' onClick={handleConfirm}>Confirm</Button>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    </Segment>
  )
}