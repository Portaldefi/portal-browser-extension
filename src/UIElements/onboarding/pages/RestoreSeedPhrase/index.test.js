import RestoreSeedPhrase from './index';
import { create, act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store'
import { Button } from 'semantic-ui-react';

describe('RestoreSeedPhraseComponent', () => {
  let initialState, mockStore, store;
  beforeEach(() => {
    initialState = {
      phrase: {
        SRF_Length: 10,
        SRF_List: [],
        password: ''
      },
      menu: {
        items: []
      }
    }
    mockStore = configureStore()
    store = mockStore(initialState)
  });

  it('should work', async () => {
    let tree;
    act(() => {
      tree = create(
        <Provider store={store}>
          <BrowserRouter>
            <RestoreSeedPhrase />
          </BrowserRouter>
        </Provider>
      );
    });

    expect(tree).toMatchSnapshot();
  });
  it('functions', async () => {
    let tree;
    act(() => {
      tree = create(
        <Provider store={store}>
          <BrowserRouter>
            <RestoreSeedPhrase />
          </BrowserRouter>
        </Provider>
      );
    });
    const buttons = tree.root.findAllByType(Button);
    act(() => buttons[0].props.onClick());
    act(() => buttons[1].props.onClick());
    expect(tree).toMatchSnapshot();
  });
  afterAll(() => jest.resetModules());
});