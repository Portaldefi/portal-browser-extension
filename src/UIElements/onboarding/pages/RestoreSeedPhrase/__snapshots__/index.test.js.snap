// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestoreSeedPhraseComponent should work 1`] = `
<div
  className="ui segment board"
>
  <div
    className="ui grid"
  >
    <div
      className="centered row"
    >
      <div
        className="ui medium header heading"
      >
        Restore Seed Phrase
      </div>
    </div>
    <div
      className="centered row"
    >
      <form
        className="ui form form-container"
        onSubmit={[Function]}
      />
    </div>
    <div
      className="row"
    >
      <p
        className="ui grey small header description extra-former-blank extra-latter-blank"
      >
        WARNING: Improper backup of the seed phrase will result in loss of funds!
      </p>
    </div>
    <div
      className="centered two column row"
    >
      <div
        className="column"
      >
        <button
          className="ui button blank-button"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="angle left icon"
            onClick={[Function]}
          />
          Back
        </button>
      </div>
      <div
        className="column"
      >
        <button
          className="ui button primary-button"
          onClick={[Function]}
        >
          Confirm
        </button>
      </div>
    </div>
  </div>
</div>
`;
